## Summary
This is my project where I demonstrate my data cleaning and transformation skills.

This dataset is taken from kaggle and is available to everyone.

## Description of the data
Dataset is a csv file that contains information about job offers for the position of a Data Scientist from the Glassdoor website.

Description of columns in this dataset:

Job Title: job title  

Salary Estimate: salary banding for each position    

Job Description: full job description

Rating: rating of each company

Company Name: name of the company  

Location: company location, city and state

Headquarters: head office location, city and state

Size: the range of the number of employees

Founded: year of foundation

Type of ownership: describes the company type i.e non-profit/public/private farm etc   

Industry: industry the company specializes in

Sector: sector the company specializes in 

Revenue: company's revenue range

Competitors: company competitors    

## Data cleaning & transformation

In [1]:
import numpy as np
import pandas as pd
import re

In [57]:
df = pd.read_csv('Uncleaned_DS_jobs.csv', index_col='index')

In [58]:
# let's look at the first 10 rows of the dataset to decide what can be changed in order to do further analysis
df.head()

Job Title               Salary Estimate  \
index                                                    
0      Sr Data Scientist  $137K-$171K (Glassdoor est.)   
1         Data Scientist  $137K-$171K (Glassdoor est.)   
2         Data Scientist  $137K-$171K (Glassdoor est.)   
3         Data Scientist  $137K-$171K (Glassdoor est.)   
4         Data Scientist  $137K-$171K (Glassdoor est.)   

                                         Job Description  Rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

                  Company Name       Location            Headquarters  \
index                                                                   
0             Healthfirst\n3.1   New York, NY            New York, NY   
1                 ManTech\n4.2  Chantilly, VA             Herndon, VA   
2          Analysis Group\n3.8     Boston, MA              Boston, MA   
3                 INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions\n2.9   New York, NY            New York, NY   

                          Size  Founded        Type of ownership  \
index                                                              
0       1001 to 5000 employees     1993   Nonprofit Organization   
1      5001 to 10000 employees     1968         Company - Public   
2       1001 to 5000 employees     1981  Private Practice / Firm   
3        501 to 1000 employees     2000         Company - Public   
4          51 to 200 employees     1998        Company - Private   

                                    Industry             Sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

                          Revenue  \
index                               
0        Unknown / Non-Applicable   
1          $1 to $2 billion (USD)   
2      $100 to $500 million (USD)   
3      $100 to $500 million (USD)   
4        Unknown / Non-Applicable   

                                             Competitors  
index                                                     
0                EmblemHealth, UnitedHealth Group, Aetna  
1                                                     -1  
2                                                     -1  
3      MKS Instruments, Pfeiffer Vacuum, Agilent Tech...  
4                   Commerce Signals, Cardlytics, Yodlee

In [4]:
# information about columns and data types
# no null values
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 672 entries, 0 to 671
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job Title          672 non-null    object 
 1   Salary Estimate    672 non-null    object 
 2   Job Description    672 non-null    object 
 3   Rating             672 non-null    float64
 4   Company Name       672 non-null    object 
 5   Location           672 non-null    object 
 6   Headquarters       672 non-null    object 
 7   Size               672 non-null    object 
 8   Founded            672 non-null    int64  
 9   Type of ownership  672 non-null    object 
 10  Industry           672 non-null    object 
 11  Sector             672 non-null    object 
 12  Revenue            672 non-null    object 
 13  Competitors        672 non-null    object 
dtypes: float64(1), int64(1), object(12)
memory usage: 78.8+ KB


### 1. Сolumn names to one view 
---

In [5]:
# make the column names more suitable
df.rename(columns=str.lower, inplace=True)

In [6]:
df.columns

Index(['job title', 'salary estimate', 'job description', 'rating',
       'company name', 'location', 'headquarters', 'size', 'founded',
       'type of ownership', 'industry', 'sector', 'revenue', 'competitors'],
      dtype='object')

In [7]:
# make the column names more suitable
df.columns = df.columns.str.replace(' ', '_')

In [8]:
df.rename(columns={'size':'staff_size'}, inplace=True)

In [9]:
df.columns

Index(['job_title', 'salary_estimate', 'job_description', 'rating',
       'company_name', 'location', 'headquarters', 'staff_size', 'founded',
       'type_of_ownership', 'industry', 'sector', 'revenue', 'competitors'],
      dtype='object')

In [10]:
# copy the dataset for further data cleaning and transformation
df_clear = df.copy()

### 2. Deleting a competitor column
---

The competitors column contains 501 rows with a "-1" value of 672 rows.

More than 70% of data about competitors is unknown, we can delete this column

In [11]:
df_clear.competitors.value_counts()

-1                                                                501
Roche, GlaxoSmithKline, Novartis                                   10
Los Alamos National Laboratory, Battelle, SRI International         6
Leidos, CACI International, Booz Allen Hamilton                     6
MIT Lincoln Laboratory, Lockheed Martin, Northrop Grumman           3
                                                                 ... 
Pfizer, GlaxoSmithKline                                             1
Square, Amazon, Apple                                               1
Lumentum Operations, Keysight Technologies, O-Net Technologies      1
Munich Re, Hannover RE, SCOR                                        1
Genomic Health, Myriad Genetics, The Broad Institute                1
Name: competitors, Length: 108, dtype: int64

In [12]:
# removed competitors column because it has 501 '-1' values
df_clear.drop(columns='competitors', inplace=True)

### 3. Removing extra values in the company name
---

In [13]:
company_name = df_clear.company_name.tolist()

In [14]:
# removed the numbers from the company name
for i in range(len(company_name)):
    company_name[i] = company_name[i].split('\n')[0]

In [15]:
df_clear.company_name = company_name

In [16]:
df_clear.head()

job_title               salary_estimate  \
index                                                    
0      Sr Data Scientist  $137K-$171K (Glassdoor est.)   
1         Data Scientist  $137K-$171K (Glassdoor est.)   
2         Data Scientist  $137K-$171K (Glassdoor est.)   
3         Data Scientist  $137K-$171K (Glassdoor est.)   
4         Data Scientist  $137K-$171K (Glassdoor est.)   

                                         job_description  rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

             company_name       location            headquarters  \
index                                                              
0             Healthfirst   New York, NY            New York, NY   
1                 ManTech  Chantilly, VA             Herndon, VA   
2          Analysis Group     Boston, MA              Boston, MA   
3                 INFICON     Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions   New York, NY            New York, NY   

                    staff_size  founded        type_of_ownership  \
index                                                              
0       1001 to 5000 employees     1993   Nonprofit Organization   
1      5001 to 10000 employees     1968         Company - Public   
2       1001 to 5000 employees     1981  Private Practice / Firm   
3        501 to 1000 employees     2000         Company - Public   
4          51 to 200 employees     1998        Company - Private   

                                    industry             sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

                          revenue  
index                              
0        Unknown / Non-Applicable  
1          $1 to $2 billion (USD)  
2      $100 to $500 million (USD)  
3      $100 to $500 million (USD)  
4        Unknown / Non-Applicable

### 4. Salary column into integer
---

In [17]:
salary_estimate = df_clear.salary_estimate.tolist()

In [18]:
# Extract salary value by removing text in parentheses
for i in range(len(salary_estimate)):
    salary_estimate[i] = salary_estimate[i].split('(')[0]

In [19]:
df_clear.salary_estimate = salary_estimate

In [20]:
df_clear.salary_estimate.unique()

array(['$137K-$171K ', '$75K-$131K ', '$79K-$131K ', '$99K-$132K ',
       '$90K-$109K ', '$101K-$165K ', '$56K-$97K ', '$79K-$106K ',
       '$71K-$123K ', '$90K-$124K ', '$91K-$150K ', '$141K-$225K ',
       '$145K-$225K', '$79K-$147K ', '$122K-$146K ', '$112K-$116K ',
       '$110K-$163K ', '$124K-$198K ', '$79K-$133K ', '$69K-$116K ',
       '$31K-$56K ', '$95K-$119K ', '$212K-$331K ', '$66K-$112K ',
       '$128K-$201K ', '$138K-$158K ', '$80K-$132K ', '$87K-$141K ',
       '$92K-$155K ', '$105K-$167K '], dtype=object)

In [21]:
df_clear.salary_estimate = df_clear.salary_estimate.str.replace(' ', '')

Creating new columns with maximum and minimum salary

In [22]:
split_data = df_clear.salary_estimate.str.split('-', expand=True)
low_salary_limit = split_data[0].tolist()
max_salary_limit = split_data[1].tolist()

In [23]:
new_idx = df_clear.columns.get_loc('salary_estimate') + 1

In [24]:
# insert the low_salary_limit column after the salary_estimate column
df_clear.insert(new_idx, 'low_salary_limit', low_salary_limit)

In [25]:
new_idx2 = df_clear.columns.get_loc('low_salary_limit') + 1

In [26]:
# insert the max_salary_limit column after the low_salary_limit column
df_clear.insert(new_idx2, 'max_salary_limit', max_salary_limit)

In [27]:
# remove extra values from the column with salary
df_clear['max_salary_limit'] = df_clear['max_salary_limit'].str.strip('$K ').astype(int)

In [28]:
df_clear['max_salary_limit'].unique()

array([171, 131, 132, 109, 165,  97, 106, 123, 124, 150, 225, 147, 146,
       116, 163, 198, 133,  56, 119, 331, 112, 201, 158, 141, 155, 167])

In [29]:
# remove extra values from the column with salary
df_clear['low_salary_limit'] = df_clear['low_salary_limit'].str.strip('$K ').astype(int)

In [30]:
df_clear['low_salary_limit'].unique()

array([137,  75,  79,  99,  90, 101,  56,  71,  91, 141, 145, 122, 112,
       110, 124,  69,  31,  95, 212,  66, 128, 138,  80,  87,  92, 105])

In [31]:
# lead to thousands
df_clear[['low_salary_limit', 'max_salary_limit']] = df_clear[['low_salary_limit', 'max_salary_limit']] * 1000

In [32]:
# remove the salary estimate column
df_clear.drop(columns='salary_estimate', inplace=True)

In [33]:
df_clear.head()

job_title  low_salary_limit  max_salary_limit  \
index                                                          
0      Sr Data Scientist            137000            171000   
1         Data Scientist            137000            171000   
2         Data Scientist            137000            171000   
3         Data Scientist            137000            171000   
4         Data Scientist            137000            171000   

                                         job_description  rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

             company_name       location            headquarters  \
index                                                              
0             Healthfirst   New York, NY            New York, NY   
1                 ManTech  Chantilly, VA             Herndon, VA   
2          Analysis Group     Boston, MA              Boston, MA   
3                 INFICON     Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions   New York, NY            New York, NY   

                    staff_size  founded        type_of_ownership  \
index                                                              
0       1001 to 5000 employees     1993   Nonprofit Organization   
1      5001 to 10000 employees     1968         Company - Public   
2       1001 to 5000 employees     1981  Private Practice / Firm   
3        501 to 1000 employees     2000         Company - Public   
4          51 to 200 employees     1998        Company - Private   

                                    industry             sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

                          revenue  
index                              
0        Unknown / Non-Applicable  
1          $1 to $2 billion (USD)  
2      $100 to $500 million (USD)  
3      $100 to $500 million (USD)  
4        Unknown / Non-Applicable

### 5. Splitting location and headquarters columns into city and state
---

Split the location column into two separate columns, where there will be information about the city and state

In [34]:
location_city = df_clear.location.str.split(', ', expand=True)[0]
location_state = df_clear.location.str.split(', ', expand=True)[1]

In [35]:
new_idx = df_clear.columns.get_loc('location') + 1

In [36]:
# insert the location_city column after the location column
df_clear.insert(new_idx, 'location_city', location_city)

In [37]:
new_idx2 = df_clear.columns.get_loc('location_city') + 1

In [38]:
# insert the location_state column after the location_city column
df_clear.insert(new_idx2, 'location_state', location_state)

Split the headquarters column into two separate columns, where there will be information about the city and state

In [39]:
headquarters_city = df_clear.headquarters.str.split(', ', expand=True)[0]
headquarters_state = df_clear.headquarters.str.split(', ', expand=True)[1]

In [40]:
new_idx = df_clear.columns.get_loc('headquarters') + 1

In [41]:
# insert the headquarters_city column after the headquarters column
df_clear.insert(new_idx, 'headquarters_city', headquarters_city)

In [42]:
new_idx2 = df_clear.columns.get_loc('headquarters_city') + 1

In [43]:
# insert the headquarters_state column after the headquarters_city column
df_clear.insert(new_idx2, 'headquarters_state', headquarters_state)

In [44]:
# delete columns location and headquarters
df_clear.drop(columns=['location', 'headquarters'], inplace=True)

In [45]:
df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 672 entries, 0 to 671
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   job_title           672 non-null    object 
 1   low_salary_limit    672 non-null    int32  
 2   max_salary_limit    672 non-null    int32  
 3   job_description     672 non-null    object 
 4   rating              672 non-null    float64
 5   company_name        672 non-null    object 
 6   location_city       672 non-null    object 
 7   location_state      649 non-null    object 
 8   headquarters_city   672 non-null    object 
 9   headquarters_state  641 non-null    object 
 10  staff_size          672 non-null    object 
 11  founded             672 non-null    int64  
 12  type_of_ownership   672 non-null    object 
 13  industry            672 non-null    object 
 14  sector              672 non-null    object 
 15  revenue             672 non-null    object 
dtypes: float

### 6. None values to the same view
---

Replace '-1', 'Unknown', 'Unknown / Non-Applicable' values with 'NO INFO'

In [46]:
df_clear.replace({'-1':'NO INFO', 'Unknown': 'NO INFO', 'Unknown / Non-Applicable': 'NO INFO'}, inplace=True)

Remove extra characters from the range of the number of employees

In [47]:
df_clear.staff_size.value_counts()

51 to 200 employees        135
1001 to 5000 employees     104
1 to 50 employees           86
201 to 500 employees        85
10000+ employees            80
501 to 1000 employees       77
5001 to 10000 employees     61
NO INFO                     44
Name: staff_size, dtype: int64

In [48]:
df_clear.staff_size = df_clear.staff_size.str.replace(' to ', '-')
df_clear.staff_size = df_clear.staff_size.str.replace(' employees', '')

Remove extra characters from company's revenue range

In [49]:
df_clear.revenue.value_counts()

NO INFO                             240
$100 to $500 million (USD)           94
$10+ billion (USD)                   63
$2 to $5 billion (USD)               45
$10 to $25 million (USD)             41
$1 to $2 billion (USD)               36
$25 to $50 million (USD)             36
$50 to $100 million (USD)            31
$1 to $5 million (USD)               31
$500 million to $1 billion (USD)     19
$5 to $10 million (USD)              14
Less than $1 million (USD)           14
$5 to $10 billion (USD)               8
Name: revenue, dtype: int64

In [50]:
df_clear.revenue = df_clear.revenue.str.replace('(USD)', '', regex=False)
df_clear.revenue = df_clear.revenue.str.replace('', '')
df_clear.revenue = df_clear.revenue.str.replace('to', '-')
df_clear.revenue = df_clear.revenue.str.replace('$', '', regex=False)

### 7. Search for keywords in the job description
---

In [51]:
# Open the file 'it_keywords.txt' in read mode using the 'open'
with open('it_keywords.txt') as file:
    keywords_list = file.read().split(',\n')

In [52]:
def skills_matches(descr):
    # Clean the description by removing unwanted characters and converting it to lowercase
    clear_descr = ' '.join([word.strip(" ,.!@#&*()-_+:;'?><") for word in descr.lower().split()])
    # Use regular expressions to find matches between the cleaned description and the list of keywords
    # Return the matched keywords as a comma-separated string
    return ', '.join([keyword for keyword in keywords_list if re.search(rf'(\s|^){re.escape(keyword)}(\s|$)', clear_descr)])

In [ ]:
# This will apply the 'skills_matches' function to each element (job description) in the 'job_description' column
# The result will be a new column 'required_skills' in the DataFrame 'df_clear'
df_clear['required_skills'] = df_clear.job_description.map(skills_matches)

In [ ]:
df_clear.head()

job_title  low_salary_limit  max_salary_limit  \
index                                                          
0      Sr Data Scientist            137000            171000   
1         Data Scientist            137000            171000   
2         Data Scientist            137000            171000   
3         Data Scientist            137000            171000   
4         Data Scientist            137000            171000   

                                         job_description  rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

             company_name location_city location_state headquarters_city  \
index                                                                      
0             Healthfirst      New York             NY          New York   
1                 ManTech     Chantilly             VA           Herndon   
2          Analysis Group        Boston             MA            Boston   
3                 INFICON        Newton             MA         Bad Ragaz   
4      Affinity Solutions      New York             NY          New York   

      headquarters_state  staff_size  founded        type_of_ownership  \
index                                                                    
0                     NY   1001-5000     1993   Nonprofit Organization   
1                     VA  5001-10000     1968         Company - Public   
2                     MA   1001-5000     1981  Private Practice / Firm   
3            Switzerland    501-1000     2000         Company - Public   
4                     NY      51-200     1998        Company - Private   

                                    industry             sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

                  revenue                                    required_skills  
index                                                                         
0                 NO INFO  machine learning, data science, data analysis,...  
1          1 - 2 billion   sql, machine learning, data science, data anal...  
2      100 - 500 million   python, r, natural language processing, machin...  
3      100 - 500 million   sql, tensorflow, pytorch, machine learning, de...  
4                 NO INFO  python, sql, r, machine learning, data science...